test/distributed/test_c10d_common.py

```diff
@@ -307,6 +313,360 @@ def _prepare_multi_device_module(
 
         return model, ddp_model, input, target
 
+    def _get_store(self):
+        return dist.FileStore(self.file_name, self.world_size)
+
+    def _get_process_group(self):
+        raise ValueError("To be implemented by child class")
+
+    def _train_model(self, model, input_var, target, loss, run_checkpoint=False, use_reentrant=True):
+        model.train()
+        if run_checkpoint:
+            output = checkpoint(model, input_var, use_reentrant=use_reentrant)
+        else:
+            output = model(input_var)
+        l = loss(output, target)
+        l.backward()
+
+    def _test_ddp_checkpointing(
+        self,
+        input_model,
+        process_group,
+        use_bucket_view,
+        find_unused_parameters=False,
+        static_graph=False,
+        run_checkpoint=False,
+        use_reentrant=True,
+        allow_none_grads=False,
+    ):
+        # to reproduce the same training results
+        torch.cuda.set_device(self.rank)
+        torch.manual_seed(31415)
+        model = copy.deepcopy(input_model).cuda()
+        ddp_model = copy.deepcopy(input_model).cuda()
+        ddp_model = nn.parallel.DistributedDataParallel(
+            ddp_model,
+            bucket_cap_mb=1,
+            gradient_as_bucket_view=use_bucket_view,
+            device_ids=[self.rank],
+            process_group=process_group,
+            find_unused_parameters=find_unused_parameters,
+            static_graph=static_graph,
+        )
+        self.assertEqual(
+            ddp_model._get_ddp_logging_data().get("static_graph", 0), static_graph
+        )
+        input, ddp_input, target, ddp_target = self._prepare_dummy_data()
+        loss = nn.MSELoss()
+        n_iters = 5
+        for i in range(n_iters):
+            model.zero_grad(set_to_none=False)
+            ddp_model.zero_grad(set_to_none=False)
+            self._train_model(model, input, target, loss, run_checkpoint=run_checkpoint, use_reentrant=use_reentrant)
+            self._train_model(
+                ddp_model, ddp_input, ddp_target, loss, run_checkpoint=run_checkpoint, use_reentrant=use_reentrant
+            )
+            for i, j in zip(model.parameters(), ddp_model.parameters()):
+                if not allow_none_grads:
+                    self.assertTrue(i.grad is not None)
+                    self.assertTrue(j.grad is not None)
+                self.assertEqual(i.grad, j.grad, rtol=1.3e-06, atol=5e-5)
```

Comment:
General question: Do we prefer `self.assertEqual()` even for `torch.Tensor`s? I have been using the new `torch.testing.assert_close()` (see [here](https://github.com/pytorch/pytorch/issues/61844)).

Reply:
I'm unsure when to use one over the other, although in this case I am directly porting the tests so don't want to change them too much.

Reply:
Just saw this and thought I clarify a little: Internally, `TestCase.assertEqual` uses the same backend as `assert_close`, but is a lot more lenient. The original plan was to remove `TestCase.assertEqual` completely in favor of `assert_close`, but we dropped it, because a lot of our tests rely on the quirks.

If you write new tests, I would encourage you to use `assert_close`, since it is more strict by default and that eases the reviewing workflow.